### DISCUSSION TITLE
Daily Interval Scheduler Misbehaving

### DISCUSSION THREAD

**Discussion by bhatrohit85** [quartznet/quartznet#1998] [Q&A]
I have a windows application in .Net 6 ( C# ) which adds Schedules using Quartz 3.6.2 . The schedules misbehave erratically , sometimes , it fires the job at same time multiple times or fires at wrong interval. we are passing UTC time and have used cron expression builder. Our requirement is simple to run a job between any start and end time during a day and also between some dates e.g

Daily between 5 am till 11 pm with interval of 19 mins starting from 20 Feb 2023 till 31 Apr 2023
Daily between 7 am till 9 pm with interval of 5 mins starting from 20 Feb 2023 till 31 Apr 2023
Daily between 12:05 pm till 11:55 pm with interval of 60 mins starting from 20 Feb 2023 till 31 Apr 2023
Daily between 9 am till 9 pm with interval of 150 mins starting from 20 Feb 2023 till 31 Apr 2023
Following is the code snippet.

var trigger = TriggerBuilder.Create()
.WithIdentity(triggerName, null)
.ForJob(jobName)
.StartAt(DateTime.SpecifyKind(scheduleInfo.DailyStartTime.Value, DateTimeKind.Utc))
.EndAt(DateTime.SpecifyKind(endDate, DateTimeKind.Utc))
.WithCronSchedule($"0 */{scheduleInfo.DailyInterval} * ? * *", action => action
.WithMisfireHandlingInstructionDoNothing().InTimeZone(TimeZoneInfo.Utc).Build())
.Build();

Please respond quickly as this issue is impacting our Client Deliverables . This has created risk and i am not getting any solution from any article . I have to get the solution asap as this is impacting timeline . Any help soon would be highly appreciated.

**Comment by jafin** [comment]
Hello,
Are you able to create a small test example that exhibits the problem? perhaps against the RAMJobstore?
Have you checked for overlapping schedules (perhaps explains jobs running at multiple times)
Have you confirmed your cron schedules are correct? perhaps validated via an online cronbuilder site?

**Comment by bhatrohit85** [comment]
> Hello, Are you able to create a small test example that exhibits the problem? perhaps against the RAMJobstore? Have you checked for overlapping schedules (perhaps explains jobs running at multiple times) Have you confirmed your cron schedules are correct? perhaps validated via an online cronbuilder site?

Hi There  , 

Thanks for replying back.

1. A complete running sample is not possible as schedulers are added from our application using Scheduler.ScheduleJob method 
using triggers and we implement Ijob interface in our application which is called by the scheduler.We are not using persistent storage but scheduler is running In-memory.

2. There are no overlapping schedules , we do keep schedule times in our db tables and then add schedulers as per the available schedules and scheduler fires the task.
3. CronExpression is simple and mentioned in the code sample above - "0 */5 * ? * *"

Is there any other information that i can provide that will help to solve the issue,

I will try to send code written in the calling application - how we add scheduler and then how we build trigger for a task,

**Comment by jafin** [comment]
>Daily between 5 am till 11 pm with interval of 19 mins starting from 20 Feb 2023 till 31 Apr 2023
Daily between 7 am till 9 pm with interval of 5 mins starting from 20 Feb 2023 till 31 Apr 2023
Daily between 12:05 pm till 11:55 pm with interval of 60 mins starting from 20 Feb 2023 till 31 Apr 2023
Daily between 9 am till 9 pm with interval of 150 mins starting from 20 Feb 2023 till 31 Apr 2023

Your provided cron expression doesn't really align with your examples?  I.e. you have a cron schedule that is every x minutes (`"0 */{scheduleInfo.DailyInterval} * ? * *"` . But there is no constraint on hours of operation or days of the month?  


`0 */5 7-21 ? 2-4 *`
would be `every 5 minutes starting at minute :00, every hour between 07am and 21pm, every month between February and April`
The month constraint and days constraint is probably pointless if you have the specific start/end dates specified as per your example, and the days are all contiguous, so perhaps just adding the hour restriction is all that is needed.

**Comment by bhatrohit85** [comment]
> > Daily between 5 am till 11 pm with interval of 19 mins starting from 20 Feb 2023 till 31 Apr 2023
> > Daily between 7 am till 9 pm with interval of 5 mins starting from 20 Feb 2023 till 31 Apr 2023
> > Daily between 12:05 pm till 11:55 pm with interval of 60 mins starting from 20 Feb 2023 till 31 Apr 2023
> > Daily between 9 am till 9 pm with interval of 150 mins starting from 20 Feb 2023 till 31 Apr 2023
> 
> Your provided cron expression doesn't really align with your examples? I.e. you have a cron schedule that is every x minutes (`"0 */{scheduleInfo.DailyInterval} * ? * *"` . But there is no constraint on hours of operation or days of the month?
> 
> `0 */5 7-21 ? 2-4 *` would be `every 5 minutes starting at minute :00, every hour between 07am and 21pm, every month between February and April` The month constraint and days constraint is probably pointless if you have the specific start/end dates specified as per your example, and the days are all contiguous, so perhaps just adding the hour restriction is all that is needed.

Hi Jafin ,

Thanks for the reply.

Earlier we were passing the hour range too. 

Do we need to pass the hour from LocalTime or can we use UTCDate also ? Also do we need to pass hour in 24 hours format ?

Regards
Rohit Bhat

**Comment by bhatrohit85** [comment]
Hi Jafin ,

I have modified the cron expression again as per your suggestion like this , i am converted date to local time and prepared expression. e,g 0 */5} 7-23 ? * * .
This caused all schedules to fire at 12 pm in the afternoon irrespective of the intervals . The same behavior occurred at 5:30 in the morning when we were not passing hour range and were specifying UTC dates

I have multiple schedules 

5 min , between 7 am to 11 pm - expression 0 */5 7-23 ? * *
9 min , between 7 am to 11 pm - expression 0 */9 7-23 ? * *
17 min , between 5 am to 11:55 pm - expression 0 */17 5-23 ? * *
**60 min , between 12:05 am to 11:55 pm** -  expression 0 */60 0-23 ? * *

**120 min ,between** **7 am to 9 pm** - need help on expression
**150 min, between** **7am to 11 pm** - need help on expression
**300 min, between** **5 am to 9 pm**  - need help on expression

Previous code for trigger with UTC 

TriggerBuilder.Create()
                                      .WithIdentity(triggerName, null)
                                      .ForJob(jobName)
                                      .StartAt(DateTime.SpecifyKind(scheduleInfo.StartDate, DateTimeKind.Utc))
                                      .EndAt(DateTime.SpecifyKind(endDate, DateTimeKind.UTC))
                                      .WithCronSchedule(scheduleInfo.CronTriggerExpression, action => action
                                      .WithMisfireHandlingInstructionDoNothing()
                                      .InTimeZone(TimeZoneInfo.Utc).Build())
                                      .Build();

Latest Code with Local TimeZone

TriggerBuilder.Create()
                                      .WithIdentity(triggerName, null)
                                      .ForJob(jobName)
                                      .StartAt(DateTime.SpecifyKind(scheduleInfo.StartDate, DateTimeKind.Local))
                                      .EndAt(DateTime.SpecifyKind(endDate, DateTimeKind.Local))
                                      .WithCronSchedule(scheduleInfo.CronTriggerExpression, action => action
                                      .WithMisfireHandlingInstructionDoNothing()
                                      .InTimeZone(TimeZoneInfo.Local).Build())
                                      .Build();

Now , i have 3 issues with latest code .

**3 schedules fired at 12 pm , service was started at 11:34 pm . Till 12 pm , all schedules were firing properly but they misfired at 12 pm and then intervals are firing properly again. This was the situation raised by the client for 5:30 in the morning when we were using UTC timezone. This is critical**

**60 min schedule not fired at all at 12:34 or even few mins after. As per cron expression , we can only put 59 in mins . So is there a way to run schedules for more than 60 mins like for 120 mins , 150 mins , 300 mins**

**So , i need to run jobs daily but with intervals of mins or hours both. If this is not possible with cron expression , is there any other triggerbuilder which can be used to simply specify starttime and endtime plus startdate and enddate.**

I am attaching the Nlog logging for more understanding. Please help me as this has been getting escalated on Prod env. It seems one issue got resolved for duplicate firing but the other one persists that 3 schedules are misfiring at 12 pm .



[2023-03-30-LogFile.log](https://github.com/quartznet/quartznet/files/11108810/2023-03-30-LogFile.log)

Regards
Rohit Bhat

**Comment by jafin** [comment]
>120 min ,between 7 am to 9 pm - need help on expression
150 min, between 7am to 11 pm - need help on expression
300 min, between 5 am to 9 pm - need help on expression

You may have hit a limitation with Quartz.NET Cron it appears it doesn't support minutes > 59.
Other cron implementations may support that.

If it were to support >59min then the expression may be (but again these appear to not work in current version of quartz.net)
```
120 min ,between 7 am to 9 pm - need help on expression
*/120 7-21 * * *

150 min, between 7am to 11 pm - need help on expression
*/150 7-23 * * *

300 min, between 5 am to 9 pm - need help on expression
*/300 5-21 * * *
```

You could try these, see if they work (multiple of hours, but because 150 isn't a multiple can't do that one:
```
Every 2 hours between 7am and 9pm
0 0 7-21/2 ? * * *

Every 5 hours between 5am - 9pm
0 0 5-21/5 ? * * *
```


I don't have any advice on why it triggers at 12pm though without further investigation/setup a test case.

**Comment by bhatrohit85** [comment]
> > 120 min ,between 7 am to 9 pm - need help on expression
> > 150 min, between 7am to 11 pm - need help on expression
> > 300 min, between 5 am to 9 pm - need help on expression
> 
> You may have hit a limitation with Quartz.NET Cron it appears it doesn't support minutes > 59. Other cron implementations may support that.
> 
> If it were to support >59min then the expression may be (but again these appear to not work in current version of quartz.net)
> 
> ```
> 120 min ,between 7 am to 9 pm - need help on expression
> */120 7-21 * * *
> 
> 150 min, between 7am to 11 pm - need help on expression
> */150 7-23 * * *
> 
> 300 min, between 5 am to 9 pm - need help on expression
> */300 5-21 * * *
> ```
> 
> You could try these, see if they work (multiple of hours, but because 150 isn't a multiple can't do that one:
> 
> ```
> Every 2 hours between 7am and 9pm
> 0 0 7-21/2 ? * * *
> 
> Every 5 hours between 5am - 9pm
> 0 0 5-21/5 ? * * *
> ```
> 
> I don't have any advice on why it triggers at 12pm though without further investigation/setup a test case.

Hi Jafin ,

I will try to provide you the sample with recreation . But can you tell if there are other trigger builder by which my requirement of schedules can be triggered as mentioned above.

Any help would be greatly appreciated. 

Regards
Rohit Bhat

**Comment by jafin** [comment]
I think there is only Simple trigger,  https://www.quartz-scheduler.net/documentation/quartz-3.x/tutorial/simpletriggers.html
But try in the discussion forums, people may have other suggestions.
